contracts/utils/packing.cairo

```diff
@@ -1,157 +1,105 @@
-func pack_rows{
-        syscall_ptr : felt*,
-        bitwise_ptr : BitwiseBuiltin*,
-        pedersen_ptr : HashBuiltin*,
-        range_check_ptr
-    }(
-        compact : felt*,
-        cell_states : felt*,
-        row : felt
-    ) -> (
-        compact : felt*
-    ):
-    if row == 0:
-        return (compact=compact)
-    end
-    alloc_locals
-    let (local compact : felt*) = pack_rows(compact=compact, cell_states=cell_states, row=row-1)
-    # (Note, on first entry, row=1 so row-1 gets the index)
-    # Create the binary encoded state for the row.
-    let (row_to_store) = pack_cols(cell_states=cell_states,
-        row=row-1, col=DIM, row_to_store=0)
-    # row_binary.write(row=row-1, value=row_to_store)
-    assert compact[row - 1] = row_to_store
-
-    return (compact=compact)
-end
 
-
-
-# Post-sim. Walk columns for a given row and saves array to state.
-func pack_cols{
+# Cells are packed according to their index
+# Starting from last cell of given array
+func pack_cells{
         syscall_ptr : felt*,
         bitwise_ptr : BitwiseBuiltin*,
         pedersen_ptr : HashBuiltin*,
         range_check_ptr
     }(
-        cell_states : felt*,
-        row : felt,
-        col : felt,
-        row_to_store : felt
+        cells_len : felt, 
+        cells : felt*, 
+        power : felt,
+        packed_cells : felt
     ) -> (
-        row_to_store : felt
+        packed_cells : felt
     ):
+    
     alloc_locals
-    if col == 0:
-        return (row_to_store)
+    
+    if cells_len == 0:
+        return (packed_cells)
     end
-    # Loops over columns, adding to a single felt using a mask.
-    let (local row_to_store) = pack_cols(cell_states=cell_states,
-        row=row, col=col-1, row_to_store=row_to_store)
-    # (Note, on first entry, col=1 so col-1 gets the index)
-
-    local pedersen_ptr : HashBuiltin* = pedersen_ptr
-    local cell_states : felt* = cell_states
-    local bitwise_ptr : BitwiseBuiltin* = bitwise_ptr
-
-
-    # Get index of cell in cell_state for this row-col combo.
-    # "Move 'row length' blocks down list, then add the column index".
-    let index = row * DIM + (col - 1)
-    let state = cell_states[index]
-
-
-    # col=0 goes in MSB. col=DIM-1 goes in LSB.
-    let binary_position = DIM - (col - 1) - 1
-    # 000...00000000011 row_to_store (old aggregator)
-    # 000...00000001000 cell_binary (cell state)
-    # 000...00000001011 bitwise OR (new aggregator)
-
-    # Binary = state * bit = state * 2**column_index
-    # E.g., For index-0: 1 * 2**0 = 0b1
-    # E.g., For index-2: 1 * 2**2 = 0b100
-
-    let (bit) = pow(2, binary_position)
-    let cell_binary = state * bit
-    # store = store OR row_binary
-    # Don't need bitwise_or here
-    #let (new_row) = bitwise_or(cell_binary, row_to_store)
-    let new_row = cell_binary + row_to_store
-
-    return (new_row)
+    
+    local bit = power / 2
+    local value = cells[cells_len - 1] * bit + packed_cells
```

Comment:
I think dividing is asking for trouble. I know you provide 2**N here, but a safer approach would be to use multiplication. Since we want to change the order we serialize (biggest first) we can do something like this:
```
   let power = power * 2
   let value = cells[cells_len - 1] * power + packed_cells
```
And you would call this function with power = 1 at the begninning.